### PR TITLE
feat: remove bloat from large files during upload (WIP)

### LINKED ISSUE #20642: [Bug]: File size is larger in payload than actual
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Description

The file size in execute request payload is larger than the original size of the file.
This is happening due to the JSON.stringify method.

### Steps To Reproduce

Try to upload a file in binary format to any plugin and verify the size of the payload. (I tried uploading to S3.)
Link to a sample file - https://drive.google.com/file/d/1Hb3VjKuzyrAdRvQCn51iK-8li6DliN_l/view?usp=sharing

### Public Sample App

_No response_

### Issue video log

_No response_

### Version

Cloud

### PR DESCRIPTION
**Pull request by ayushpahwa**
## Description

Currently, we try to upload large files by converting their binaries into strings which leads to bloat in size. This is because converting to bytes in a multi-byte encoding usually takes a larger space and white characters are also included. We were also doing multiple modifications which were just adding to the bloat. 

Hence, we are now converting the binary data into an array buffer to prevent this. This buffer is added to the multi-part form data request as a new part and we add a pointer in the pace of the data which used to be present earlier. This allows us to have minimal bloat on the payload while sending the request.

TLDR: fix for uploading large files by changing the data type used for upload.

*TODO:*
- [x] Client side payload changes
- [x] Server side double escape logic fixes
- [x] Server side tests
- [x] Server side refactor
- [ ] Cypress tests

Fixes #20642 

Media

## Type of change

- New feature (non-breaking change which adds functionality)

## How Has This Been Tested?

- Manual

### Dev activity
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [ ] PR is being merged under a feature flag

### QA activity:
- [ ] Test plan has been approved by relevant developers
- [ ] Test plan has been peer reviewed by QA
- [ ] Cypress test cases have been added and approved by either SDET or manual QA
- [ ] Organized project review call with relevant stakeholders after Round 1/2 of QA
- [ ] Added Test Plan Approved label after reveiwing all Cypress test

### CODE REVIEW

File: app/client/src/sagas/ActionExecution/PluginActionSaga.ts [hunk 1]
```diff
@@ -184,7 +196,14 @@ function* readBlob(blobUrl: string): any {
     if (fileType === FileDataTypes.Base64) {
       reader.readAsDataURL(file);
     } else if (fileType === FileDataTypes.Binary) {
-      reader.readAsBinaryString(file);
+      //check size of the file, if less than 5mb, go with binary string method
+      // else go with array buffer method
+      if (file.size < FILE_SIZE_LIMIT_FOR_BLOBS) {
+        // TODO: this method is deprecated, use readAsText instead
+        reader.readAsBinaryString(file);
```
Comment (ChandanBalajiBP) [inline app/client/src/sagas/ActionExecution/PluginActionSaga.ts:203]: What is the cost involved in updating this method? Can we do this now?

File: app/client/src/sagas/ActionExecution/PluginActionSaga.ts [hunk 2]
```diff
@@ -184,7 +196,14 @@ function* readBlob(blobUrl: string): any {
     if (fileType === FileDataTypes.Base64) {
       reader.readAsDataURL(file);
     } else if (fileType === FileDataTypes.Binary) {
-      reader.readAsBinaryString(file);
+      //check size of the file, if less than 5mb, go with binary string method
```
Comment (ChandanBalajiBP) [inline app/client/src/sagas/ActionExecution/PluginActionSaga.ts:199]: Can you add a comment to give context on why this condition and method was added.

File: app/client/src/sagas/ActionExecution/PluginActionSaga.ts [hunk 3]
```diff
@@ -316,18 +346,52 @@ function* evaluateActionParams(
       value = yield call(resolvingBlobUrls, value, executeActionRequest, i);
     }
 
+    let useBlobMaps = false;
+
     if (typeof value === "object") {
+      // This is used in cases of large files
+      if (value.hasOwnProperty("blobUrlPaths")) {
```
Comment (ChandanBalajiBP) [inline app/client/src/sagas/ActionExecution/PluginActionSaga.ts:353]: Can you add a comment on the reason behind this action.

File: app/server/appsmith-server/src/main/java/com/appsmith/server/services/ce/NewActionServiceCEImpl.java [hunk 4]
```diff
@@ -135,6 +148,11 @@ public class NewActionServiceCEImpl extends BaseService<NewActionRepository, New
     public static final PluginType JS_PLUGIN_TYPE = PluginType.JS;
     public static final String JS_PLUGIN_PACKAGE_NAME = "js-plugin";
 
+    static final String PARAM_KEY_REGEX = "k\\d+";
+    static final String BLOB_KEY_REGEX = "blob:[0-9a-fA-F]{8}\\b-[0-9a-fA-F]{4}\\b-[0-9a-fA-F]{4}\\b-[0-9a-fA-F]{4}\\b-[0-9a-fA-F]{12}";
```
[the remark this answers is not in the document]
Comment (nidhi-nair) [reply app/server/appsmith-server/src/main/java/com/appsmith/server/services/ce/NewActionServiceCEImpl.java:152]: Yeah this is intentional, because we're not sure what kind of usage users may have inside their bindings, for example they might have a tertiary operator to choose between file data.

[the remark this answers is not in the document]
Comment (subrata71) [reply app/server/appsmith-server/src/main/java/com/appsmith/server/services/ce/NewActionServiceCEImpl.java:152]: > for example they might have a tertiary operator to choose between file data.

How does it affect the `part` name?

**Review by subrata71 — DISMISSED** [review]

**Comment by AnandShahs123** [comment]
@ayushpahwa @nidhi-nair - A file which is uploaded through filepicker widget is failing to upload through REST API in both binary and base64 format. 
payload too large

File: app/client/src/sagas/ActionExecution/PluginActionSaga.ts (hunk 1 above)
Comment (ayushpahwa) [reply app/client/src/sagas/ActionExecution/PluginActionSaga.ts:203]: There are some changes in encoding and it was leading to an error. Will need to deep dive into this, hence added a todo here.

File: app/client/src/sagas/ActionExecution/PluginActionSaga.ts [hunk 5]
```diff
@@ -321,17 +374,51 @@ function* evaluateActionParams(
     }
 
     if (typeof value === "object") {
+      // This is used in cases of large files, we store the bloburls with the path they were set in
+      // This helps in creating a unique map of blob urls to blob data when passing to the server
+      if (value.hasOwnProperty("blobUrlPaths")) {
```
Comment (ChandanBalajiBP) [inline app/client/src/sagas/ActionExecution/PluginActionSaga.ts:379]: As you have same set of codes, can we move this to a function?

File: app/client/src/sagas/ActionExecution/PluginActionSaga.ts [hunk 6]
```diff
@@ -321,17 +374,51 @@ function* evaluateActionParams(
...
+          },
+        );
+        unset(value, "blobUrlPaths");
+      }
+
       value = JSON.stringify(value);
     }
 
-    value = new Blob([value], { type: "text/plain" });
+    if (!useBlobMaps) {
```
Comment (ChandanBalajiBP) [inline app/client/src/sagas/ActionExecution/PluginActionSaga.ts:399]: Can you add comments

File: app/client/src/sagas/ActionExecution/PluginActionSaga.ts [hunk 7]
```diff
@@ -321,17 +374,51 @@ function* evaluateActionParams(
...
+      value = new Blob([value], { type: "text/plain" });
+    }
     bindingsMap[key] = `k${i}`;
     bindingBlob.push({ name: `k${i}`, value: value });
+    const paramProperties = executeActionRequest.paramProperties[`k${i}`];
```
Comment (ChandanBalajiBP) [inline app/client/src/sagas/ActionExecution/PluginActionSaga.ts:404]: Can you add comments

File: app/client/src/sagas/ActionExecution/PluginActionSaga.ts [hunk 8]
```diff
@@ -235,6 +258,12 @@ function* resolvingBlobUrls(
       const blobUrl = value[blobUrlPath] as string;
       const resolvedBlobValue: unknown = yield call(readBlob, blobUrl);
       set(value, blobUrlPath, resolvedBlobValue);
+      const blobUrlPaths = get(value, "blobUrlPaths", {}) as Record<
+        string,
+        string
+      >;
+      set(blobUrlPaths, blobUrlPath, blobUrl);
```
Comment (ChandanBalajiBP) [inline app/client/src/sagas/ActionExecution/PluginActionSaga.ts:265]: Can you add comments

File: app/server/appsmith-server/src/main/java/com/appsmith/server/services/ce/NewActionServiceCEImpl.java [hunk 9]
```diff
@@ -207,6 +227,11 @@ public NewActionServiceCEImpl(Scheduler scheduler,
         this.applicationPermission = applicationPermission;
         this.pagePermission = pagePermission;
         this.actionPermission = actionPermission;
+
+        this.patternList.add(Pattern.compile(PARAM_KEY_REGEX));
+        this.patternList.add(Pattern.compile(BLOB_KEY_REGEX));
+        this.patternList.add(Pattern.compile(EXECUTE_ACTION_DTO));
```
Comment (sondermanish) [inline app/server/appsmith-server/src/main/java/com/appsmith/server/services/ce/NewActionServiceCEImpl.java:233]: Do we need to iterate over Execute action and parameterMap as well?

File: app/client/src/sagas/ActionExecution/PluginActionSaga.ts (hunk 6 above)
Comment (ayushpahwa) [reply app/client/src/sagas/ActionExecution/PluginActionSaga.ts:423]: Added

File: app/client/src/sagas/ActionExecution/PluginActionSaga.ts (hunk 8 above)
Comment (ayushpahwa) [reply app/client/src/sagas/ActionExecution/PluginActionSaga.ts]: Added

**Review by sondermanish — APPROVED** [review]

**Review by ChandanBalajiBP — APPROVED** [review]
LGTM

**Review by ChandanBalajiBP — APPROVED** [review]
